File: tests/c/sigmask_preemption.c
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

#define _GNU_SOURCE

#include <pthread.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

enum { THREADS = 4, ROUNDS = 500 };

struct worker {
  unsigned id;
  uint64_t checksum;
  int error;
};

static void *run_worker(void *opaque) {
  struct worker *worker = opaque;
  sigset_t blocked;
  sigset_t previous;
  sigemptyset(&blocked);
  sigaddset(&blocked, SIGUSR1);

  uint64_t value = worker->id + 1;
  for (unsigned round = 0; round < ROUNDS; ++round) {
    int error = pthread_sigmask(SIG_BLOCK, &blocked, &previous);
    if (error == 0) {
      error = pthread_sigmask(SIG_SETMASK, &previous, NULL);
    }
    if (error != 0) {
      worker->error = error;
      return NULL;
    }
    value = value * 6364136223846793005ULL + 1442695040888963407ULL;
  }
  worker->checksum = value;
  return NULL;
}

int main(void) {
  pthread_t threads[THREADS];
  struct worker workers[THREADS] = {0};
  uint64_t checksum = 0;

  for (unsigned index = 0; index < THREADS; ++index) {
    workers[index].id = index;
    if (pthread_create(&threads[index], NULL, run_worker, &workers[index]) != 0) {
      return 10;
    }
  }
  for (unsigned index = 0; index < THREADS; ++index) {
    if (pthread_join(threads[index], NULL) != 0 || workers[index].error != 0) {
      fprintf(stderr, "worker=%u error=%d\n", index, workers[index].error);
      return 20;
    }
    checksum ^= workers[index].checksum;
  }

  printf("sigmask-stress threads=%d rounds=%d checksum=%016lx\n", THREADS,
         ROUNDS, (unsigned long)checksum);
  return EXIT_SUCCESS;
}

File: hermit-cli/tests/sigmask_preemption.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::fs;
use std::path::Path;
use std::process::Command;

const EXPECTED_STDOUT: &str = "sigmask-stress threads=4 rounds=500 checksum=f183820163e0384c\n";

fn compile_guest() -> std::path::PathBuf {
    let repository = Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("hermit-cli should be inside the repository");
    let build_root = Path::new(env!("CARGO_TARGET_TMPDIR")).join("sigmask-preemption");
    fs::create_dir_all(&build_root).expect("failed to create sigmask build directory");
    let binary = build_root.join("sigmask_preemption");
    let output = Command::new("cc")
        .args([
            "-std=c11", "-O2", "-g", "-Wall", "-Wextra", "-Werror", "-pthread",
        ])
        .arg(repository.join("tests/c/sigmask_preemption.c"))
        .arg("-o")
        .arg(&binary)
        .output()
        .expect("failed to compile sigmask preemption guest");
    assert!(
        output.status.success(),
        "sigmask guest compilation failed:\n{}",
        String::from_utf8_lossy(&output.stderr),
    );
    binary
}

#[test]
fn interrupted_signal_mask_injections_are_retried() {
    let binary = compile_guest();
    let program = binary.to_str().expect("guest path should be UTF-8");

    let strict = Command::new("timeout")
        .args(["--kill-after", "10s", "90s"])
        .arg(env!("CARGO_BIN_EXE_hermit"))
        .args(["run", "--strict", "--max-timeslice=1000", "--", program])
        .output()
        .expect("failed to run strict sigmask preemption guest");
    assert!(
        strict.status.success(),
        "strict sigmask guest failed:\nstdout:\n{}\nstderr:\n{}",
        String::from_utf8_lossy(&strict.stdout),
        String::from_utf8_lossy(&strict.stderr),
    );
    assert_eq!(String::from_utf8_lossy(&strict.stdout), EXPECTED_STDOUT);

    let verified = Command::new("timeout")
        .args(["--kill-after", "10s", "90s"])
        .arg(env!("CARGO_BIN_EXE_hermit"))
        .args([
            "--log=info",
            "run",
            "--strict",
            "--verify",
            "--max-timeslice=1000",
            "--",
            program,
        ])
        .output()
        .expect("failed to verify sigmask preemption guest");
    assert!(
        verified.status.success(),
        "sigmask verification failed:\nstdout:\n{}\nstderr:\n{}",
        String::from_utf8_lossy(&verified.stdout),
        String::from_utf8_lossy(&verified.stderr),
    );
    assert!(
        String::from_utf8_lossy(&verified.stderr).contains("Determinism verified"),
        "sigmask verification omitted success marker:\n{}",
        String::from_utf8_lossy(&verified.stderr),
    );
}
